Signal/src/call/PeerConnectionClient.swift

```diff
@@ -334,181 +377,220 @@ class PeerConnectionClient: NSObject, RTCPeerConnectionDelegate, RTCDataChannelD
-                        let error = OWSErrorMakeUnableToProcessServerResponseError()
-                        reject(error)
-                        return
-                    }
+        PeerConnectionClient.signalingQueue.async { [weak self] in
+            guard let strongSelf = handle.get() else { return }
+            strongSelf.assertOnSignalingQueue()
 
-                    let hardenedSessionDescription = HardenedRTCSessionDescription(rtcSessionDescription: sessionDescription)
+            guard let peerConnection = strongSelf.peerConnection else {
+                Logger.debug("\(strongSelf.logTag) \(#function) Ignoring obsolete event in terminated client")
+                return
+            }
 
-                    self.setLocalSessionDescriptionInternal(hardenedSessionDescription)
-                        .then(on: PeerConnectionClient.signalingQueue) {
-                            fulfill(hardenedSessionDescription)
-                        }.catch { error in
-                            reject(error)
-                    }
+            Logger.debug("\(strongSelf.logTag) negotiating answer session.")
+
+            peerConnection.answer(for: constraints, completionHandler: { (sdp: RTCSessionDescription?, error: Error?) in
+                PeerConnectionClient.signalingQueue.async { [weak self] in
+                    completion(sdp, error)
                 }
             })
         }
+        return promise
     }
 
     public func addRemoteIceCandidate(_ candidate: RTCIceCandidate) {
-        PeerConnectionClient.signalingQueue.async {
-            guard self.peerConnection != nil else {
-                Logger.debug("\(self.TAG) \(#function) Ignoring obsolete event in terminated client")
+        let handle = self.handle
+        PeerConnectionClient.signalingQueue.async { [weak self] in
+            guard let strongSelf = handle.get() else { return }
+            guard let peerConnection = strongSelf.peerConnection else {
+                Logger.debug("\(strongSelf.logTag) \(#function) Ignoring obsolete event in terminated client")
                 return
             }
-            Logger.info("\(self.TAG) adding remote ICE candidate: \(candidate.sdp)")
-            self.peerConnection.add(candidate)
+            Logger.info("\(strongSelf.logTag) adding remote ICE candidate: \(candidate.sdp)")
+            peerConnection.add(candidate)
         }
     }
 
     public func terminate() {
         SwiftAssertIsOnMainThread(#function)
-        Logger.debug("\(TAG) in \(#function)")
+        Logger.debug("\(logTag) in \(#function)")
 
         // Clear the delegate immediately so that we can guarantee that
         // no delegate methods are called after terminate() returns.
         delegate = nil
 
+        // Clear the handle immediately so that enqueued work is aborted
+        // going forward.
+        handle.clear()
```

Comment:
During teardown, we call `terminate` which calls `terminateInternal` on the signaling queue.  We eagerly clear the handle.